rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+providing a playground for developers to try out new ideas and potentially deliver performance
+quickly.
+
+While all the important APIs are implemented in C++ for efficiency, every part of the system can be
+easily switched to customized python implementation. For example,
+
+**Customize design space in python**. Can be a python function that does the schedule
+
+```python
+def schedule_matmul(sch) -> sch:
+    i, j, k = sch.get_loops(sch.get_block("matmul"))
+    i_tiles = sch.sample_perfect_tile(i, n=4)
+    j_tiles = sch.sample_perfect_tile(j, n=4)
+    k_tiles = sch.sample_perfect_tile(k, n=2)
+    # Tile the loops according to the random variables
+    i_0, i_1, i_2, i_3 = sch.split(loop=i, factors=i_tiles)
+    j_0, j_1, j_2, j_3 = sch.split(loop=j, factors=j_tiles)
+    k_0, k_1 = sch.split(loop=k, factors=k_tiles)
+    # Organize the loops into "SSRSRS" 6-level tiles
+    sch.reorder(
+        i_0, j_0, # S
+        i_1, j_1, # S
+        k_0,      # R
+        i_2, j_2, # S
+        k_1,      # R
+        i_3, j_3, # S
+    )
+    return sch
+```
+
+**Customize composite schedule in python**. We provide two ways to define a composite schedule in
+python:
+
+Method 1. Derive from `PyCompositeSchedule`, and implement two methods `initialize` and `apply`:
+
+```python
+class MultiLevelTiling(PyCompositeSchedule):
+    def initialize(...):
+        # initialize the class, usually this method is empty
+        ...
+
+    def apply(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+        # write any python code, including:
+        # - analyze `block`
+        # - invoke schedule primitives in `sch`
+        # - do debug printing
+        ...
+```
+
+Method 2. A decorator as the syntactic sugar if the `initialize` method is empty, which converts the
+function to the `apply` method.
+
+```python
+@tir.as_composite_schedule(name="multi-level-tiling")
+def multi_level_tiling(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+    # write any python code, including:
+    # - analyze `block`
+    # - invoke schedule primitives in `sch`
+    # - do debug printing
+    ...
+```
+
+**Customize exploration strategies in python**. Developers can implement any search algorithm in
+python as well by deriving from `PySearchPolicy`, and the syntax is identical to customizing with
+`PyCompositeSchedule`.
+
+**Other customizable components**. This list includes:
+
+* Cost model
+* Database
+* Measure callbacks
+* Feature extractor
+* Program builder & runner
+* Analysis methods
+* ...
+
+In a short summary, almost every component of the system is decoupled with each other and extensions
+could be easily plugged in.
+
+## 5. Drawbacks
+
+We are not aware of any drawbacks of the proposed system.
+
+## 6. Rationale and alternatives
+
+The system is designed with the principle of minimalism: different from alternative solutions, we do
```

Comment:
explain...

Reply:
The paragraph is not well written. Just revised it:

```markdown
The system is designed with the principle of minimalism: Assuming users already know TensorIR
scheduling APIs, there are no more extra API set to learn, and the previous TensorIR scheduling
programs will work out of box with meta schedule. It could potentially lower the bar of adopting
this system.
```